### PR TITLE
Try resolving issues discovered with -Wall -Werror

### PR DESCRIPTION
**Pull request by tepperly**
I tried switching to C++ 17 and turning on `-Wall -Werror` during compile. This found a variety of issues.

It's hard for me as an outside to know when unused variables can be deleted versus just marking them as ``[[maybe_unused]]`. The key here is whether calculating the variable value has important side effects.

The main changes here involve:
* ensuring that constructors initialize member variables in the order they're declared because that's the order that C++ will do them
* fixing comparisons between `int` and `size_t` (unsigned long) in conditionals
* removing or marking as `[[maybe_unused]]` variables that are never used
* a few situations where `=` was used instead of `==` inside an assert
* a destructor that needs to be `virtual`

### CODE REVIEW

File: src/Interface/hiopInterfacePrimalDecomp.hpp [hunk 1]
```diff
@@ -142,7 +142,7 @@ class hiopInterfacePriDecProblem
                             const hiopVector& x0,
                             const std::string& mem_space);
 
-    ~RecourseApproxEvaluator();
+    virtual ~RecourseApproxEvaluator();
```
Comment (tepperly) [inline src/Interface/hiopInterfacePrimalDecomp.hpp:145]: IMHO, this is an important change to ensure that the right destructor is called when `delete evaluator` is called from `hiopAlgPrimalDecomp.cpp`.

File: src/LinAlg/hiopMatrixSparseCSRSeq.cpp [hunk 2]
```diff
@@ -1064,7 +1056,7 @@ void hiopMatrixSparseCSRSeq::form_transpose_from_symbolic(const hiopMatrixSparse
     }
     //here row_starts_(==w) contains the row starts
   }
-  assert(irowptr_[nrows_] = nnz_);
+  assert(irowptr_[nrows_] == nnz_);
```
Comment (tepperly) [inline src/LinAlg/hiopMatrixSparseCSRSeq.cpp:1059]: Assignment to equality check!

**Comment by nychiang** [comment]
For those [[maybe_unused]] variables, should I correct them in this PR? Or do you prefer me merging this PR first and then I correct those issues in another PR? @tepperly

**Comment by tepperly** [comment]
> For those [[maybe_unused]] variables, should I correct them in this PR? Or do you prefer me merging this PR first and then I correct those issues in another PR? @tepperly

Which ever you prefer. It's up to you. I had to make my changes on a fork of your repository. If you wanted to fix in this PR, I think you'll need to push your changes to [my fork of your repo](https://github.com/tepperly/hiop).

**Comment by tepperly** [comment]
Is there something I can do to facilitate the review and merging of this PR?

**Comment by cnpetra** [comment]
> Is there something I can do to facilitate the review and merging of this PR?

it is in the works and on our to-do list for this week.

**Comment by nychiang** [comment]
@tepperly, I just created one PR in your folk, to correct some additional files due to the issue discovered with `-Wall -Werror`.  I have no problem to compile the code on my MAC, running with RAJA/MPI/HSL. All the unit test work fine.

However, I wonder if you could work on this PR in the LLNL folk, i.e., move `tepperly:feature/headers-wall-werror-proof` to `LLNL:feature/headers-wall-werror-proof` and then file this PR. If you file PR inside the LLNL folk, I believe all the pipeline test (CI) will start in LLNL and PNNL clusters.

I am not sure why all the CI tests do not work while your PR is initialized from an external folk.

**Comment by nychiang** [comment]
@tepperly  With @cameronrutherford's help, the PNNL CI is working now. Once you file this PR within LLNL domain, we can get feedbacks from all the CI tests.

**Comment by tepperly** [comment]
> @tepperly With @cameronrutherford's help, the PNNL CI is working now. Once you file this PR within LLNL domain, we can get feedbacks from all the CI tests.

I think you need to give user @tepperly access to push to your repo. I forked because I can't push to a branch on `git@github.com:LLNL/hiop.git`.

```sh
$ git push origin feature/headers-wall-werror-proof
ERROR: Permission to LLNL/hiop.git denied to tepperly.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by cameronrutherford** [comment]
https://gitlab.pnnl.gov/exasgd/frameworks/exago/-/merge_requests/488 - recently did this in ExaGO FWIW so should be reasonably easy to add this to HiOp pipelines as well if you want

**Comment by nychiang** [comment]
@cnpetra Let's close this one. We filed another PR inside LLNL domain, to use the CI features. See #653

**Pull request closed without merging**